File: contrib/gateway-peers/terminal/src/agentm_terminal/frontends/tui/router.py
"""Wire-frame router: ``metadata.kind`` -> one widget mutation.

The single place wire shapes meet widgets. The gateway streams the full event
taxonomy (see ``wire_driver``); this maps each ``metadata.kind`` to a transcript
mutation, a StatusBar update, or a toast. Unknown kinds are ignored (forward-
compatible). See ``.claude/designs/textual-tui.md`` §5.4.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import TYPE_CHECKING, Any

from .widgets import (
    ApprovalBlock,
    AssistantTurn,
    SubagentBlock,
    SystemTurn,
    ToolBlock,
)

if TYPE_CHECKING:
    from .app import AgentMTui

Handler = Callable[[dict[str, Any], dict[str, Any]], Awaitable[None]]

# Control/observability kinds shown as a toast, with the formatter + variant.
# Self-modification (agent-triggered reload) is rendered louder on purpose.
_SELF_MODIFY_TRIGGERS = {"agent", "propose_change_approved"}


class Router:
    def __init__(self, app: AgentMTui) -> None:
        self._app = app
        self._active: AssistantTurn | None = None
        self._tools: dict[str, ToolBlock] = {}
        self._children: dict[str, SubagentBlock] = {}
        self._dispatch: dict[str, Handler] = {
            "turn_start": self._turn_start,
            "stream_text": self._stream_text,
            "stream_thinking": self._stream_thinking,
            "tool_call": self._tool_call,
            "tool_result": self._tool_result,
            "usage": self._usage,
            "child_start": self._child_start,
            "child_end": self._child_end,
            "assistant_text": self._assistant_text,
            "approval_request": self._approval_request,
            "approval_resolved": self._approval_resolved,
            "diagnostic_warning": self._diagnostic,
            "diagnostic_error": self._diagnostic,
            "agent_end": self._agent_end,
            "api_send_user_message": self._injected,
            "session_ready": self._session_ready,
            "api_register": self._api_register,
            "cost_budget_exceeded": self._budget,
            "extension_install": self._control,
            "extension_reload": self._control,
            "extension_unload": self._control,
            "resource_write": self._control,
            "plan_submitted": self._control,
            "after_compact": self._control,
            "command_dispatched": self._control,
        }

    @property
    def active(self) -> AssistantTurn | None:
        return self._active

    async def dispatch(self, body: dict[str, Any]) -> None:
        meta = body.get("metadata")
        meta = meta if isinstance(meta, dict) else {}
        kind = str(meta.get("kind") or "assistant_text")
        fn = self._dispatch.get(kind)
        if fn is not None:
            await fn(body, meta)

    # --- conversation ---------------------------------------------------

    async def _ensure_turn(self) -> AssistantTurn:
        if self._active is None:
            turn = AssistantTurn()
            await self._app.mount_widget(turn)
            self._active = turn
        return self._active

    async def _turn_start(self, body: dict, meta: dict) -> None:
        turn = AssistantTurn()
        await self._app.mount_widget(turn)
        self._active = turn
        self._app.set_phase("thinking")

    async def _stream_text(self, body: dict, meta: dict) -> None:
        (await self._ensure_turn()).append_text(str(body.get("content") or ""))
        self._app.set_phase("streaming")

    async def _stream_thinking(self, body: dict, meta: dict) -> None:
        (await self._ensure_turn()).append_thinking(str(body.get("content") or ""))
        self._app.set_phase("thinking")

    async def _tool_call(self, body: dict, meta: dict) -> None:
        turn = await self._ensure_turn()
        args = meta.get("args")
        block = ToolBlock(
            name=str(meta.get("name") or "tool"),
            args=args if isinstance(args, dict) else {},
        )
        self._tools[str(meta.get("tool_call_id") or "")] = block
        await turn.add_block(block)
        self._app.set_phase("tool")
        self._app.scroll_end()

    async def _tool_result(self, body: dict, meta: dict) -> None:
        block = self._tools.get(str(meta.get("tool_call_id") or ""))
        if block is not None:
            block.set_result(
                ok=bool(meta.get("ok", True)), content=str(body.get("content") or "")
            )
            self._app.scroll_end()

    async def _usage(self, body: dict, meta: dict) -> None:
        self._app.status.tokens_in += int(meta.get("input_tokens") or 0)
        self._app.status.tokens_out += int(meta.get("output_tokens") or 0)
        self._app.show_status()

    async def _child_start(self, body: dict, meta: dict) -> None:
        turn = await self._ensure_turn()
        block = SubagentBlock(purpose=str(meta.get("purpose") or "subagent"))
        self._children[str(meta.get("child_id") or "")] = block
        await turn.add_block(block)
        self._app.set_phase("subagent")
        self._app.scroll_end()

    async def _child_end(self, body: dict, meta: dict) -> None:
        block = self._children.get(str(meta.get("child_id") or ""))
        if block is not None:
            block.finish(error=meta.get("error"))

    async def _assistant_text(self, body: dict, meta: dict) -> None:
        turn = await self._ensure_turn()
        content = str(body.get("content") or "")
        await turn.set_final_text(content)
        self._app.note_assistant_text(content)  # for /copy-last
        self._active = None  # the turn is complete; the next turn_start opens one
        self._app.set_phase("idle")
        self._app.scroll_end()

    async def _agent_end(self, body: dict, meta: dict) -> None:
        self._app.mark_idle()

    # --- approvals ------------------------------------------------------

    async def _approval_request(self, body: dict, meta: dict) -> None:
        buttons = [b for b in (body.get("buttons") or []) if isinstance(b, dict)]
        block = ApprovalBlock(
            content=str(body.get("content") or ""),
            buttons=buttons,
            on_select=self._app.send_button,
        )
        await self._app.mount_widget(block)
        block.focus()
        self._app.scroll_end()

    async def _approval_resolved(self, body: dict, meta: dict) -> None:
        self._app.toast(f"approval {meta.get('decision', 'resolved')}")

    # --- control / observability ---------------------------------------

    async def _diagnostic(self, body: dict, meta: dict) -> None:
        self._app.toast(str(body.get("content") or ""), variant="warn")

    async def _injected(self, body: dict, meta: dict) -> None:
        await self._app.mount_widget(
            SystemTurn(
                str(body.get("content") or ""),
                source=str(meta.get("extension") or "extension"),
            )
        )
        self._app.scroll_end()

    async def _session_ready(self, body: dict, meta: dict) -> None:
        self._app.status.model = str(meta.get("model") or self._app.status.model)
        tools = meta.get("tool_names")
        if isinstance(tools, list):
            for t in tools:
                self._app.catalog.add_tool(str(t))
            self._app.status.tool_count = len(self._app.catalog.tools)
        commands = meta.get("command_names")
        if isinstance(commands, list):
            for c in commands:
                self._app.catalog.add_command(_as_slash(str(c)))
        self._app.show_status()

    async def _api_register(self, body: dict, meta: dict) -> None:
        name = str(meta.get("name") or "")
        if meta.get("reg_kind") == "tool":
            self._app.catalog.add_tool(name)
            self._app.status.tool_count = len(self._app.catalog.tools)
            self._app.show_status()
        elif meta.get("reg_kind") == "command":
            self._app.catalog.add_command(_as_slash(name))

    async def _budget(self, body: dict, meta: dict) -> None:
        self._app.status.budget_exceeded = True
        self._app.catalog.budget = (
            f"exceeded: {meta.get('used')}/{meta.get('limit')} "
            f"{meta.get('currency', '')}".strip()
        )
        self._app.show_status()
        self._app.toast(
            f"budget exceeded: {meta.get('used')}/{meta.get('limit')} "
            f"{meta.get('currency', '')}".strip(),
            variant="warn",
        )

    async def _control(self, body: dict, meta: dict) -> None:
        kind = str(meta.get("kind") or "")
        # Record extension lifecycle into the catalog (backs /extensions),
        # for every phase — not just the ones we toast.
        if kind == "extension_install":
            mod = str(meta.get("module_path") or "")
            err = meta.get("error")
            self._app.catalog.extensions[mod] = (
                f"{meta.get('phase')}: {err}" if err else str(meta.get("phase") or "")
            )
        elif kind in ("extension_reload", "extension_unload"):
            self._app.catalog.extensions[str(meta.get("name") or "")] = kind.split(
                "_", 1
            )[1]
        # Skip the noisy bootstrap install start/end; surface failures + reloads.
        if kind == "extension_install" and meta.get("phase") != "error":
            return
        text, variant = _control_summary(kind, meta)
        if text:
            self._app.toast(text, variant=variant)


def _as_slash(name: str) -> str:
    """Normalise a command name to its ``/name`` invocation form."""
    return name if name.startswith("/") else f"/{name}"


def _control_summary(kind: str, meta: dict[str, Any]) -> tuple[str, str]:
    if kind == "extension_reload":
        self_modify = bool(meta.get("is_self_modify")) or (
            meta.get("trigger") in _SELF_MODIFY_TRIGGERS
        )
        prefix = "★ self-modify" if self_modify else "reload"
        err = f"  ({meta.get('error')})" if meta.get("error") else ""
        return f"{prefix}: {meta.get('name')}{err}", (
            "selfmod" if self_modify else "info"
        )
    if kind == "extension_unload":
        return f"unload: {meta.get('name')}", "info"
    if kind == "extension_install":  # only reaches here on error
        return f"install failed: {meta.get('module_path')}", "warn"
    if kind == "resource_write":
        return f"wrote {meta.get('path')}  ({meta.get('author')})", "info"
    if kind == "plan_submitted":
        return "plan submitted", "info"
    if kind == "after_compact":
        return (
            f"compacted: kept {meta.get('kept')}, dropped {meta.get('discarded')}",
            "info",
        )
    if kind == "command_dispatched":
        return f"/{meta.get('name')}", "info"
    return "", "info"


__all__ = ["Router"]

File: contrib/gateway-peers/terminal-go/internal/app/router.go
package app

import (
	"fmt"
	"time"

	"github.com/AoyangSpace/agentm-terminal/internal/blocks"
	"github.com/AoyangSpace/agentm-terminal/internal/theme"
)

// Router dispatches wire outbound events to Model state mutations.
type Router struct{}

// Dispatch routes a wire event body to the appropriate handler based on
// metadata.kind. Unknown kinds are silently ignored (forward-compatible).
func (r *Router) Dispatch(m *Model, body map[string]any) {
	if body == nil {
		return
	}
	meta, _ := body["metadata"].(map[string]any)
	kind := "assistant_text"
	if meta != nil {
		if k, ok := meta["kind"].(string); ok && k != "" {
			kind = k
		}
	}

	switch kind {
	case "turn_start":
		r.turnStart(m, body, meta)
	case "stream_text":
		r.streamText(m, body, meta)
	case "stream_thinking":
		r.streamThinking(m, body, meta)
	case "tool_call":
		r.toolCall(m, body, meta)
	case "tool_result":
		r.toolResult(m, body, meta)
	case "assistant_text":
		r.assistantText(m, body, meta)
	case "agent_end":
		r.agentEnd(m)
	case "usage":
		r.usage(m, body, meta)
	case "session_ready":
		r.sessionReady(m, body, meta)
	case "approval_request":
		r.approvalRequest(m, body, meta)
	case "approval_resolved":
		r.approvalResolved(m, body, meta)
	case "diagnostic_warning":
		r.diagnostic(m, body, "warn")
	case "diagnostic_error":
		r.diagnostic(m, body, "warn")
	case "child_start":
		r.childStart(m, body, meta)
	case "child_end":
		r.childEnd(m, body, meta)
	case "cost_budget_exceeded":
		r.budgetExceeded(m, body, meta)
	case "extension_install":
		if phase, _ := meta["phase"].(string); phase == "error" {
			r.diagnostic(m, body, "warn")
		}
	case "extension_reload":
		r.extensionReload(m, body, meta)
	case "api_register":
		r.apiRegister(m, body, meta)
	case "after_compact":
		r.afterCompact(m, body, meta)
	}
}

// ensureActiveTurn guarantees an active assistant turn exists.
func (r *Router) ensureActiveTurn(m *Model) *blocks.AssistantTurn {
	if m.activeTurn == nil {
		m.activeTurn = &blocks.AssistantTurn{GlamourStyle: m.glamourStyle}
		m.transcript = append(m.transcript, m.activeTurn)
	}
	return m.activeTurn
}

func (r *Router) turnStart(m *Model, _ map[string]any, _ map[string]any) {
	turn := &blocks.AssistantTurn{GlamourStyle: m.glamourStyle}
	m.transcript = append(m.transcript, turn)
	m.activeTurn = turn
	m.turnStartTime = time.Now()
	sm := m.status.GetModel()
	sm.Phase = theme.PhaseThinking
	m.status.Update(sm)
}

func (r *Router) streamText(m *Model, body map[string]any, _ map[string]any) {
	turn := r.ensureActiveTurn(m)
	if content, ok := body["content"].(string); ok {
		if turn.OpenText() == nil {
			tb := &blocks.TextBlock{GlamourStyle: m.glamourStyle}
			turn.AppendSegment(tb)
			turn.SetOpenText(tb)
			turn.SetOpenThinking(nil)
		}
		turn.OpenText().Text += content
	}
	sm := m.status.GetModel()
	sm.Phase = theme.PhaseStreaming
	m.status.Update(sm)
}

func (r *Router) streamThinking(m *Model, body map[string]any, _ map[string]any) {
	turn := r.ensureActiveTurn(m)
	if content, ok := body["content"].(string); ok {
		if turn.OpenThinking() == nil {
			tb := blocks.NewThinkingBlock()
			turn.AppendSegment(tb)
			turn.SetOpenThinking(tb)
			turn.SetOpenText(nil)
		}
		turn.OpenThinking().Text += content
	}
	sm := m.status.GetModel()
	sm.Phase = theme.PhaseThinking
	m.status.Update(sm)
}

func (r *Router) toolCall(m *Model, _ map[string]any, meta map[string]any) {
	turn := r.ensureActiveTurn(m)
	name, _ := meta["name"].(string)
	if name == "" {
		name = "unknown"
	}

	var args map[string]any
	if a, ok := meta["args"].(map[string]any); ok {
		args = a
	}

	tb := blocks.NewToolBlock(name, args)
	turn.AppendSegment(tb)
	// Close any open streaming segments so the next think/text starts fresh.
	turn.SetOpenText(nil)
	turn.SetOpenThinking(nil)

	if callID, ok := meta["tool_call_id"].(string); ok && callID != "" {
		m.toolRegistry[callID] = tb
	}

	sm := m.status.GetModel()
	sm.Phase = theme.PhaseTool
	sm.ToolCount++
	m.status.Update(sm)
}

func (r *Router) toolResult(m *Model, body map[string]any, meta map[string]any) {
	callID, _ := meta["tool_call_id"].(string)
	tb := m.toolRegistry[callID]
	if tb == nil {
		return
	}
	if result, ok := body["content"].(string); ok {
		tb.Result = result
	}
	if ok, exists := meta["ok"].(bool); exists {
		tb.OK = ok
	} else {
		tb.OK = true
	}
	tb.Done = true
}

func (r *Router) assistantText(m *Model, body map[string]any, _ map[string]any) {
	turn := r.ensureActiveTurn(m)
	if content, ok := body["content"].(string); ok {
		// Set the content on the last TextBlock, or create one if none exists.
		tb := turn.LastTextBlock()
		if tb == nil {
			tb = &blocks.TextBlock{GlamourStyle: m.glamourStyle}
			turn.AppendSegment(tb)
		}
		tb.Text = content
	}
	turn.SetComplete()
	m.activeTurn = nil
	sm := m.status.GetModel()
	sm.Phase = theme.PhaseIdle
	m.status.Update(sm)
}

func (r *Router) agentEnd(m *Model) {
	m.inFlight = false
	m.activeTurn = nil
	sm := m.status.GetModel()
	sm.Phase = theme.PhaseIdle
	m.status.Update(sm)
}

func (r *Router) usage(m *Model, body map[string]any, _ map[string]any) {
	sm := m.status.GetModel()
	if tokIn, ok := toInt(body["input_tokens"]); ok {
		sm.TokensIn += tokIn
	}
	if tokOut, ok := toInt(body["output_tokens"]); ok {
		sm.TokensOut += tokOut
	}
	if cost, ok := toFloat(body["cost"]); ok {
		sm.CostTurn = cost
	}
	if sessionCost, ok := toFloat(body["session_cost"]); ok {
		sm.CostSession = sessionCost
	}
	elapsed := time.Since(m.turnStartTime)
	if elapsed.Seconds() > 0 && sm.TokensOut > 0 {
		sm.TokPerSec = float64(sm.TokensOut) / elapsed.Seconds()
	}
	m.status.Update(sm)
}

func (r *Router) sessionReady(m *Model, body map[string]any, meta map[string]any) {
	if model, ok := meta["model"].(string); ok {
		sm := m.status.GetModel()
		sm.Model = model
		m.status.Update(sm)
	}
	if ctxWin, ok := toInt(meta["context_window"]); ok {
		sm := m.status.GetModel()
		sm.CtxTotal = ctxWin
		m.status.Update(sm)
	}
	if tools, ok := body["tools"].([]any); ok {
		for _, t := range tools {
			if name, ok := t.(string); ok {
				m.addToolToCatalog(name)
			}
		}
	}
	if cmds, ok := body["commands"].([]any); ok {
		for _, c := range cmds {
			if name, ok := c.(string); ok {
				m.addCommand(name)
			}
		}
	}
}

func (r *Router) approvalRequest(m *Model, body map[string]any, meta map[string]any) {
	content, _ := body["content"].(string)
	if content == "" {
		content = "Approval required"
	}

	var buttons []blocks.Button
	if rawBtns, ok := meta["buttons"].([]any); ok {
		for _, rb := range rawBtns {
			if bm, ok := rb.(map[string]any); ok {
				label, _ := bm["label"].(string)
				value, _ := bm["value"].(string)
				style, _ := bm["style"].(string)
				buttons = append(buttons, blocks.Button{Label: label, Value: value, Style: style})
			}
		}
	}
	if len(buttons) == 0 {
		buttons = []blocks.Button{
			{Label: "Allow", Value: "allow", Style: "primary"},
			{Label: "Deny", Value: "deny", Style: "danger"},
		}
	}

	ab := blocks.NewApprovalBlock(content, buttons)
	ab.SetCollapsed(false)

	if turn := m.activeTurn; turn != nil {
		turn.Approvals = append(turn.Approvals, ab)
	} else {
		// Standalone approval as a system block
		m.transcript = append(m.transcript, &blocks.SystemTurn{
			Content: content,
			Source:  "approval",
		})
	}
	m.pendingApproval = ab
}

func (r *Router) approvalResolved(m *Model, body map[string]any, _ map[string]any) {
	decision, _ := body["decision"].(string)
	if decision == "" {
		decision = "resolved"
	}
	m.toasts.Push("approval: "+decision, "info", 3*time.Second)
	m.pendingApproval = nil
}

func (r *Router) diagnostic(m *Model, body map[string]any, variant string) {
	content, _ := body["content"].(string)
	if content == "" {
		content = "diagnostic event"
	}
	m.toasts.Push(content, variant, 5*time.Second)
}

func (r *Router) childStart(m *Model, _ map[string]any, meta map[string]any) {
	turn := r.ensureActiveTurn(m)
	purpose, _ := meta["purpose"].(string)
	childID, _ := meta["child_id"].(string)

	sb := &blocks.SubagentBlock{Purpose: purpose}
	turn.Children = append(turn.Children, sb)
	if childID != "" {
		m.childRegistry[childID] = sb
	}

	sm := m.status.GetModel()
	sm.Phase = theme.PhaseSubagent
	m.status.Update(sm)
}

func (r *Router) childEnd(m *Model, _ map[string]any, meta map[string]any) {
	childID, _ := meta["child_id"].(string)
	if sb, ok := m.childRegistry[childID]; ok {
		sb.Done = true
		if errMsg, ok := meta["error"].(string); ok {
			sb.Error = errMsg
		}
	}
}

func (r *Router) budgetExceeded(m *Model, body map[string]any, _ map[string]any) {
	cost, _ := toFloat(body["cost"])
	sm := m.status.GetModel()
	sm.CostSession = cost
	sm.BudgetWarn = true
	m.status.Update(sm)
	m.toasts.Push("cost budget exceeded", "warn", 10*time.Second)
}

func (r *Router) extensionReload(m *Model, body map[string]any, meta map[string]any) {
	isSelfMod, _ := meta["is_self_modify"].(bool)
	name, _ := body["extension"].(string)
	variant := "info"
	msg := fmt.Sprintf("extension reloaded: %s", name)
	if isSelfMod {
		variant = "selfmod"
		msg = fmt.Sprintf("self-modification: %s reloaded", name)
	}
	m.toasts.Push(msg, variant, 5*time.Second)
}

func (r *Router) apiRegister(m *Model, body map[string]any, meta map[string]any) {
	regType, _ := meta["type"].(string)
	name, _ := body["name"].(string)
	switch regType {
	case "tool":
		m.addToolToCatalog(name)
	case "command":
		m.addCommand(name)
	}
}

func (r *Router) afterCompact(m *Model, body map[string]any, _ map[string]any) {
	m.toasts.Push("context compacted", "info", 3*time.Second)
	if ctxUsed, ok := toInt(body["ctx_used"]); ok {
		sm := m.status.GetModel()
		sm.CtxUsed = ctxUsed
		m.status.Update(sm)
	}
}

// toInt coerces a JSON number (float64) or int to int.
func toInt(v any) (int, bool) {
	switch n := v.(type) {
	case float64:
		return int(n), true
	case int:
		return n, true
	case int64:
		return int(n), true
	default:
		return 0, false
	}
}

// toFloat coerces a JSON number to float64.
func toFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	default:
		return 0, false
	}
}
